src/middlewared/middlewared/plugins/pool_/dataset_details.py

```diff
@@ -0,0 +1,221 @@
+
+        return info
+
+    @private
+    def collapse_datasets(self, dataset, collapsed):
+        collapsed.append(dataset)
+        for child in dataset.get('children', []):
+            self.collapse_datasets(child, collapsed)
+
+    @private
+    def build_iscsi_share_info(self):
+        t_to_e = self.middleware.call_sync('iscsi.targetextent.query')
+        t = {i['id']: i for i in self.middleware.call_sync('iscsi.target.query')}
+        e = {i['id']: i for i in self.middleware.call_sync('iscsi.extent.query')}
+
+        iscsi_shares = []
+        for i in filter(lambda x: x['target'] in t and t[x['target']]['groups'] and x['extent'] in e, t_to_e):
+            """
+            1. make sure target's and extent's id exist in the target to extent table
+            2. make sure the target has `groups` entry since, without it, it's impossible
+                that it's being shared via iscsi
+            """
+            iscsi_shares.append({'extent': e[i['extent']], 'target': t[i['target']]})
+
+        return iscsi_shares
+
+    @private
+    def get_snapcount_and_encryption_status(self, ds, mntinfo):
+        snap_count = 0
+        locked = False
+        if ds['type'] == 'FILESYSTEM':
+            # FIXME: determining zvol snapshot count requires iterating over
+            # all snapshots for a given zvol and then counting them which is
+            # painful. This will be moot when this is merged upstream
+            # https://github.com/openzfs/zfs/pull/13635
+
+            try:
+                st = os.stat(f'{ds["mountpoint"]}/.zfs/snapshot')
+            except FileNotFoundError:
+                # means that the zfs snapshot dir doesn't exist which
+                # can only happen if the dataset is encrypted and "locked"
+                # (unmounted) or the dataset is unmounted
+                locked = ds['encrypted']
+            else:
+                if st.st_ino == ZFSCTL.INO_SNAPDIR:
+                    # dataset isn't locked and the inode of the snapshot dir
+                    # is what we expect so we're able to determine the snapshot
+                    # count by reading st_nlink
+                    snap_count = st.st_nlink - 2  # (remove 2 for `.` and `..`)
+                else:
+                    # zfs snapshot dir exists but inode doesn't match reality
+                    # which can only happen if dataset is unmounted and the
+                    # same directory structure is created manually
+                    locked = ds['encrypted']
+
+        return snap_count, locked
+
+    @private
+    def get_nfs_shares(self, ds, nfsshares, mntinfo):
+        nfs_shares = []
+        found = False
+        for nfsshare in nfsshares:
+            if nfsshare['path'] == ds['mountpoint']:
+                # the share path is the actual dataset
+                found = True
+            else:
+                try:
+                    devid = os.stat(nfsshare['path']).st_dev
```

Comment:
You do `os.stat` for every NFS share path N times (where N = len(collapsed), can it make sense to cache it? Same for other share types.

Reply:
Good point. I'll cache devid for the path and only stat if it doesn't exist.